lib/c-api/src/wasm_c_api/unstable/middlewares/metering.rs

```diff
@@ -0,0 +1,300 @@
+//!         // Oh, it failed!
+//!         assert(trap != NULL);
+//!
+//!         // There is 0 point left… they are exhausted.
+//!         assert(wasmer_metering_get_remaining_points(instance) == 0);
+//!     }
+//!     
+//!     wasm_instance_delete(instance);
+//!     wasm_module_delete(module);
+//!     wasm_store_delete(store);
+//!     wasm_engine_delete(engine);
+//!
+//!     return 0;
+//! }
+//! #    })
+//! #    .success();
+//! # }
+//! ```
+
+use super::super::super::instance::wasm_instance_t;
+use super::super::parser::operator::wasmer_parser_operator_t;
+use super::wasmer_middleware_t;
+use std::sync::Arc;
+use wasmer::wasmparser::Operator;
+use wasmer_middlewares::{
+    metering::{get_remaining_points, set_remaining_points, MeteringPoints},
+    Metering,
+};
+
+/// Opaque type representing a metering middleware.
+///
+/// To transform this specific middleware into a generic one, please
+/// see [`wasmer_metering_as_middleware`].
+///
+/// # Example
+///
+/// See module's documentation.
+#[allow(non_camel_case_types)]
+pub struct wasmer_metering_t {
+    pub(crate) inner: Arc<Metering<Box<dyn Fn(&Operator) -> u64 + Send + Sync>>>,
+}
+
+/// Function type to represent a user-defined cost function
+/// implemented in C.
+///
+/// # Example
+///
+/// See module's documentation.
+#[allow(non_camel_case_types)]
+pub type wasmer_metering_cost_function_t = extern "C" fn(operator: wasmer_parser_operator_t) -> u64;
+
+/// Creates a new metering middleware with an initial limit, i.e. a
+/// total number of operators to execute (regarding their respective
+/// cost), in addition to a cost function. The cost function defines
+/// the cost of an operation, that will decrease the initial limit.
+///
+/// # Example
+///
+/// See module's documentation.
+#[no_mangle]
+pub extern "C" fn wasmer_metering_new(
+    initial_limit: u64,
+    cost_function: wasmer_metering_cost_function_t,
+) -> Box<wasmer_metering_t> {
+    let cost_function = move |operator: &Operator| -> u64 { cost_function(operator.into()) };
+
+    Box::new(wasmer_metering_t {
+        inner: Arc::new(Metering::new(initial_limit, Box::new(cost_function))),
+    })
+}
+
+/// Deletes a [`wasmer_metering_t`].
+///
+/// # Example
+///
+/// See module's documentation.
+#[no_mangle]
+pub extern "C" fn wasmer_metering_delete(_metering: Option<Box<wasmer_metering_t>>) {}
+
+/// Returns the remaining metering points. Zero means points are
+/// exhausted, otherwise it returns the number of points.
+///
+/// # Example
+///
+/// See module's documentation.
+#[no_mangle]
+pub extern "C" fn wasmer_metering_get_remaining_points(instance: &wasm_instance_t) -> u64 {
+    match get_remaining_points(&instance.inner) {
+        MeteringPoints::Remaining(value) => value,
+        MeteringPoints::Exhausted => 0,
+    }
+}
+
+/// Set a new amount of points for the given metering middleware.
+///
+/// # Example
+///
+/// This example is pointless as the number of points aren't updated
```

Comment:
Is that a TODO then?

Otherwise it's not pointless, it's just that its point is only to illustrate the `wasmer_metering_set_remaining_points` function.

Reply:
:+1: Updated